pkg/adhoc/server/server.go

```diff
@@ -0,0 +1,119 @@
+package server
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/sirupsen/logrus"
+
+	"github.com/pyroscope-io/pyroscope/pkg/adhoc/util"
+)
+
+type profile struct {
+	ID        string    `json:"id"`
+	Name      string    `json:"name"`
+	UpdatedAt time.Time `json:"updated_at"`
+}
+
+type server struct {
+	log      logrus.FieldLogger
+	enabled  bool
+	profiles map[string]profile
+	mtx      sync.RWMutex
+}
+
+type Server interface {
+	AddRoutes(router *mux.Router) http.HandlerFunc
+}
+
+func New(log logrus.FieldLogger, enabled bool) Server {
+	return &server{log: log, enabled: enabled, profiles: make(map[string]profile)}
+}
+
+func (s *server) AddRoutes(r *mux.Router) http.HandlerFunc {
+	if s.enabled {
+		r.HandleFunc("/v1/profiles", s.Profiles)
+		r.HandleFunc("/v1/profile/{id:[0-9a-f]+}", s.Profile)
+		r.HandleFunc("/v1/diff/{id1[0-9a-f]+}/{id2[0-9a-f]+}", s.Diff)
+	}
+	return r.ServeHTTP
+}
+
+// Profiles retrieves the list of profiles for the local pyroscope data directory.
+// The profiles are assigned a unique ID (hash based) which is then used for retrieval.
+// This requires a bit of extra work to setup the IDs but prevents
+// the clients from accesing the filesystem directly, removing that whole attack vector.
+//
+// The profiles are retrieved every time the endpoint is requested,
+// which should be good enough as massive access to this auth endpoint is not expected.
+func (s *server) Profiles(w http.ResponseWriter, _ *http.Request) {
+	profiles := make(map[string]profile, 0)
+	err := filepath.Walk(util.DataDirectory(), func(path string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.Mode().IsRegular() {
+			id := fmt.Sprintf("%x", sha256.Sum256([]byte(info.Name())))
+			if p, ok := profiles[id]; ok {
+				return fmt.Errorf("A hash collision detected between %s and %s, please report it", info.Name(), p.Name)
+			}
+			profiles[id] = profile{ID: id, Name: info.Name(), UpdatedAt: info.ModTime()}
+		}
+		return nil
+	})
+	if err != nil {
+		s.log.WithError(err).Error("Unable to retrieve the profile list")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	s.mtx.Lock()
+	s.profiles = profiles
+	s.mtx.Unlock()
+	w.Header().Set("Content-Type", "application/json")
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	if err := json.NewEncoder(w).Encode(s.profiles); err != nil {
+		s.log.WithError(err).Error("Unable to encode the profile list")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
```

Comment:
Maybe we should split looking up files and sending responding

Reply:
What do you mean?

Reply:
I am suggesting a more MVC approach, splitting the controller and the service, instead of doing everything (business logic and response sending) directly in the handler.

Reply:
Ah, yes. There's so little logic right now that it's a bit overkill, but as it grows we may want to separate the business logic from the formatting :+1: